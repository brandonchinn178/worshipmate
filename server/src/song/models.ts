export type Song = {
  id: number
  slug: string
  title: string
  artist: string
  recommendedKey: string
  timeSignature: TimeSignature
  bpm: number
}

export type TimeSignature = [number, number]

export type SearchFilters = {
  recommendedKey?: string
  bpm?: number
  timeSignature?: TimeSignature
}

export type Artist = {
  id: number
  slug: string
  name: string
}
