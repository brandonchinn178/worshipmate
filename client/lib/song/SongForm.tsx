import * as _ from 'lodash'
import { Controller, useForm } from 'react-hook-form'
import Select from 'react-select'
import styled from 'styled-components'

import { color } from '~/theme'
import { ErrorMessage } from '~/ui-kit/ErrorMessage'

import { TimeSignature } from './models'

export type SongFormValues = {
  slug: string
  title: string
  artist: string
  recommendedKey: string
  timeSignature: TimeSignature
  bpm: number
  themes: string[]
}

type SongFormProps = {
  initialSong?: SongFormValues
  onSubmit: (payload: SongFormValues) => Promise<void>
}

export function SongForm({ initialSong, onSubmit }: SongFormProps) {
  const {
    formState: { isSubmitting, isSubmitSuccessful, errors },
    control,
    handleSubmit,
    register,
  } = useForm<SongFormValues>({
    defaultValues: {
      slug: initialSong?.slug,
      title: initialSong?.title,
      artist: initialSong?.artist,
      recommendedKey: initialSong?.recommendedKey,
      timeSignature: initialSong?.timeSignature ?? [4, 4],
      bpm: initialSong?.bpm,
      themes: initialSong?.themes ?? [],
    },
  })

  const postSubmitLoading = isSubmitting || isSubmitSuccessful

  const timeSignatureOptions = getTimeSignatureOptions(
    initialSong?.timeSignature,
  )

  return (
    <FormContainer onSubmit={handleSubmit(onSubmit)}>
      {initialSong && (
        <FormField>
          <label htmlFor="slug">Slug</label>
          <input id="slug" {...register('slug', { required: true })} />
          <ErrorMessage name="slug" errors={errors} />
        </FormField>
      )}

      <FormField>
        <label htmlFor="title">Title</label>
        <input id="title" {...register('title', { required: true })} />
        <ErrorMessage name="title" errors={errors} />
      </FormField>

      <FormField>
        <label htmlFor="artist">Artist</label>
        <input id="artist" {...register('artist', { required: true })} />
        <ErrorMessage name="artist" errors={errors} />
      </FormField>

      <FormField>
        <label htmlFor="recommendedKey">Recommended Key</label>
        <input
          id="recommendedKey"
          {...register('recommendedKey', { required: true })}
        />
        <ErrorMessage name="recommendedKey" errors={errors} />
      </FormField>

      <FormField>
        <label htmlFor="timeSignature">Time Signature</label>
        <Controller
          name="timeSignature"
          control={control}
          render={({ field: { onChange, onBlur, value } }) => {
            return (
              <Select
                instanceId="timeSignature"
                inputId="timeSignature"
                options={timeSignatureOptions}
                value={_.find(timeSignatureOptions, { value })}
                onChange={(option) => onChange(option?.value)}
                onBlur={onBlur}
                className="time-signature-container"
              />
            )
          }}
        />
      </FormField>

      <FormField>
        <label htmlFor="bpm">BPM</label>
        <input
          id="bpm"
          type="number"
          {...register('bpm', { required: true, valueAsNumber: true })}
        />
        <ErrorMessage name="bpm" errors={errors} />
      </FormField>

      <button disabled={postSubmitLoading}>Submit</button>
    </FormContainer>
  )
}

const getTimeSignatureOptions = (existingTimeSignature?: TimeSignature) => {
  const timeSignatures = _.compact([
    existingTimeSignature,
    [4, 4],
    [3, 4],
    [6, 8],
    [2, 4],
  ])

  const options = _.map(timeSignatures, (value) => {
    const [top, bottom] = value
    return {
      label: `${top}/${bottom}`,
      value,
    }
  })

  return _.uniqBy(options, 'label')
}

const FormContainer = styled.form`
  display: grid;
  width: 600px;
  margin: 50px auto 0;

  border: 4px ridge ${color('lightGray')};
  padding: 25px 75px;
`

const FormField = styled.div`
  display: grid;
  grid-template-columns: 200px 1fr;
  grid-template-areas:
    'label input'
    'error error';
  align-items: center;

  margin-bottom: 1rem;

  ${ErrorMessage.Style} {
    grid-area: error;
    line-height: 1.5rem;
    margin-bottom: -0.5rem;
  }
`
