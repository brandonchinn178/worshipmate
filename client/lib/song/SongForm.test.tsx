import { screen, waitFor } from '@testing-library/react'
import userEvent from '@testing-library/user-event'

import { renderUI, waitForAllStateChanges } from '~jest-utils'

import { TimeSignature } from './models'
import { SongForm } from './SongForm'

describe('initial render', () => {
  it('renders an empty form', () => {
    const { container } = renderUI(<SongForm onSubmit={jest.fn()} />)

    expect(container).toMatchSnapshot()

    const inputTitle = screen.getByLabelText('Title')
    expect(inputTitle).toBeVisible()
    expect(inputTitle).toHaveValue('')

    const inputArtist = screen.getByLabelText('Artist')
    expect(inputArtist).toBeVisible()
    expect(inputArtist).toHaveValue('')

    const inputRecommendedKey = screen.getByLabelText('Recommended Key')
    expect(inputRecommendedKey).toBeVisible()
    expect(inputRecommendedKey).toHaveValue('')

    const selectTimeSignature = screen.getByLabelText('Time Signature')
    expect(selectTimeSignature).toBeVisible()
    expect(selectTimeSignature).toHaveValue('[4,4]')

    const inputBpm = screen.getByLabelText('BPM')
    expect(inputBpm).toBeVisible()
    expect(inputBpm).toHaveValue(null)
  })

  it('renders an initially populated form', () => {
    const song = {
      slug: 'blessed-be-your-name',
      title: 'Blessed Be Your Name',
      artist: 'Matt Redman',
      recommendedKey: 'A',
      timeSignature: [3, 4] as TimeSignature,
      bpm: 120,
    }
    renderUI(<SongForm initialSong={song} onSubmit={jest.fn()} />)

    expect(screen.getByLabelText('Title')).toHaveValue(song.title)
    expect(screen.getByLabelText('Artist')).toHaveValue(song.artist)
    expect(screen.getByLabelText('Recommended Key')).toHaveValue(
      song.recommendedKey,
    )
    expect(screen.getByLabelText('Time Signature')).toHaveValue('[3,4]')
    expect(screen.getByLabelText('BPM')).toHaveValue(120)

    // verify that '3/4' isn't duplicated
    expect(screen.getAllByText('3/4')).toHaveLength(1)
  })

  it('renders populated form with non-standard time signature', () => {
    const song = {
      slug: 'take-five',
      title: 'Take Five',
      artist: 'Dave Brubeck',
      recommendedKey: 'Eb',
      timeSignature: [5, 4] as TimeSignature,
      bpm: 170,
    }
    renderUI(<SongForm initialSong={song} onSubmit={jest.fn()} />)

    const selectTimeSignature = screen.getByLabelText('Time Signature')
    expect(selectTimeSignature).toHaveValue('[5,4]')
  })
})

describe('form validation', () => {
  it('triggers onSubmit when passes', async () => {
    const onSubmit = jest.fn()
    renderUI(<SongForm onSubmit={onSubmit} />)

    populateForm()

    await waitFor(() => {
      expect(onSubmit).toHaveBeenCalled()
    })
  })

  it('does not trigger onSubmit with empty fields', async () => {
    const onSubmit = jest.fn()
    renderUI(<SongForm onSubmit={onSubmit} />)

    userEvent.click(screen.getByRole('button'))

    const errorMessage = await screen.findByText('Title is required')
    expect(errorMessage).toBeVisible()
    expect(onSubmit).not.toHaveBeenCalled()
  })
})

describe('submit button', () => {
  it('is enabled at start', async () => {
    renderUI(<SongForm onSubmit={jest.fn()} />)

    expect(screen.getByRole('button')).not.toHaveAttribute('disabled')
  })

  it('is disabled during and after submitting', async () => {
    const onSubmit = jest.fn()
    renderUI(<SongForm onSubmit={onSubmit} />)

    populateForm()

    expect(screen.getByRole('button')).toHaveAttribute('disabled')

    await waitFor(() => {
      expect(onSubmit).toHaveBeenCalled()
    })

    expect(screen.getByRole('button')).toHaveAttribute('disabled')
  })

  it('is enabled after form validation fails', async () => {
    const onSubmit = jest.fn()
    renderUI(<SongForm onSubmit={onSubmit} />)

    userEvent.click(screen.getByRole('button'))

    await waitForAllStateChanges()
    expect(onSubmit).not.toHaveBeenCalled()

    expect(screen.getByRole('button')).not.toHaveAttribute('disabled')
  })

  it('is enabled after submission fails', async () => {
    const onSubmit = jest.fn().mockRejectedValue(new Error())

    renderUI(<SongForm onSubmit={onSubmit} />)

    populateForm()

    expect(screen.getByRole('button')).toHaveAttribute('disabled')

    await waitFor(() => {
      expect(onSubmit).toHaveBeenCalled()
    })

    expect(screen.getByRole('button')).not.toHaveAttribute('disabled')
  })
})

const populateForm = () => {
  userEvent.type(screen.getByLabelText('Title'), 'Blessed Be Your Name')
  userEvent.type(screen.getByLabelText('Artist'), 'Matt Redman')
  userEvent.type(screen.getByLabelText('Recommended Key'), 'A')
  userEvent.selectOptions(screen.getByLabelText('Time Signature'), '[3,4]')
  userEvent.type(screen.getByLabelText('BPM'), '120')
  userEvent.click(screen.getByRole('button'))
}
