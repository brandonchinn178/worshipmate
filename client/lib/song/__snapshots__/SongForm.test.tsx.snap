// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`initial render renders an empty form 1`] = `
<div>
  <form
    class="SongForm__FormContainer-iup8lu-0 fZvGnO"
  >
    <div
      class="SongForm__FormField-iup8lu-1 eTJHAx"
    >
      <label
        for="title"
      >
        Title
      </label>
      <input
        id="title"
        name="title"
      />
    </div>
    <div
      class="SongForm__FormField-iup8lu-1 eTJHAx"
    >
      <label
        for="artist"
      >
        Artist
      </label>
      <input
        id="artist"
        name="artist"
      />
    </div>
    <div
      class="SongForm__FormField-iup8lu-1 eTJHAx"
    >
      <label
        for="recommendedKey"
      >
        Recommended Key
      </label>
      <input
        id="recommendedKey"
        name="recommendedKey"
      />
    </div>
    <div
      class="SongForm__FormField-iup8lu-1 eTJHAx"
    >
      <label
        for="timeSignature"
      >
        Time Signature
      </label>
      <select
        id="timeSignature"
      >
        <option
          value="[4,4]"
        >
          4/4
        </option>
        <option
          value="[3,4]"
        >
          3/4
        </option>
        <option
          value="[6,8]"
        >
          6/8
        </option>
        <option
          value="[2,4]"
        >
          2/4
        </option>
      </select>
    </div>
    <div
      class="SongForm__FormField-iup8lu-1 eTJHAx"
    >
      <label
        for="bpm"
      >
        BPM
      </label>
      <input
        id="bpm"
        name="bpm"
        type="number"
      />
    </div>
    <button>
      Submit
    </button>
  </form>
</div>
`;
